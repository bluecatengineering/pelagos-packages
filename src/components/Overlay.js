import {cloneElement, Component} from 'react';
import {createPortal, findDOMNode} from 'react-dom';
import PropTypes from 'prop-types';
import pick from 'lodash-es/pick';

import './Overlay.less';

const PROPS = ['positionLeft', 'positionTop', 'arrowOffsetLeft', 'arrowOffsetTop'];

const getOffset = (position, size, windowSize) =>
	position < 0 ? -position : position + size > windowSize ? windowSize - position - size : 0;

const addClassName = (child) => (child.props.className ? child.props.className + ' ' : '') + 'Overlay';

export default class Overlay extends Component {
	static propTypes = {
		show: PropTypes.bool,
		placement: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
		target: PropTypes.any,
		children: PropTypes.element,
	};

	state = {
		positionLeft: 0,
		positionTop: 0,
		arrowOffsetLeft: null,
		arrowOffsetTop: null,
		visible: false,
	};

	componentDidMount() {
		if (this.props.show) {
			this.updatePosition();
		}
	}

	componentWillUnmount() {
		if (this.timer) {
			clearTimeout(this.timer);
		}
	}

	componentDidUpdate(prevProps, prevState) {
		if (this.props.show && (!prevProps.show || (!prevProps.target && this.props.target))) {
			this.updatePosition();
		}

		if (this.state.visible) {
			if (!prevState.visible) {
				this.getNode().classList.add('Overlay--visible');
			} else if (!this.props.show) {
				this.getNode().classList.remove('Overlay--visible');
				this.timer = setTimeout(() => ((this.timer = null), this.setState({visible: false})), 300);
			}
		}
	}

	getNode() {
		// The child is dynamic, there's no other way to get the node
		// eslint-disable-next-line react/no-find-dom-node
		return findDOMNode(this);
	}

	updatePosition() {
		const target = this.props.target;
		if (target) {
			// eslint-disable-next-line react/no-find-dom-node
			const rt = findDOMNode(target).getBoundingClientRect();
			const {width: childWidth, height: childHeight} = this.getNode().getBoundingClientRect();
			const placement = this.props.placement;

			let positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop;
			if (placement === 'left' || placement === 'right') {
				positionTop = rt.top + (rt.height - childHeight) / 2;
				positionLeft = rt.left + (placement === 'left' ? -childWidth : rt.width);
				const offset = getOffset(positionTop, childHeight, window.innerHeight);
				positionTop += offset;
				arrowOffsetTop = 50 * (1 - (2 * offset) / childHeight) + '%';
			} else {
				positionLeft = rt.left + (rt.width - childWidth) / 2;
				positionTop = rt.top + (placement === 'top' ? -childHeight : rt.height);
				const offset = getOffset(positionLeft, childWidth, window.innerWidth);
				positionLeft += offset;
				arrowOffsetLeft = 50 * (1 - (2 * offset) / childWidth) + '%';
			}
			this.setState({positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop, visible: true});
		}
	}

	render() {
		return this.props.show || this.state.visible
			? createPortal(
					cloneElement(this.props.children, {
						...pick(this.state, PROPS),
						className: addClassName(this.props.children),
						placement: this.props.placement,
					}),
					document.body
			  )
			: null;
	}
}
