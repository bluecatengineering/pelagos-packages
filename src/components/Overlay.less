.Overlay {
	opacity: 0;
	transition: opacity 0.3s ease-out;
}

.Overlay--visible {
	opacity: 1;
}
