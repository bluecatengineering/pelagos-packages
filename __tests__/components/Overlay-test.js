import React from 'react';
import ReactDOM from 'react-dom';
import {mount, shallow} from 'enzyme';

import Overlay from '../../src/components/Overlay';

jest.unmock('../../src/components/Overlay');

const Test = () => <div />;

describe('Overlay', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<Overlay show={true} placement="left">
					<Test />
				</Overlay>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when show is false', () => {
			const wrapper = shallow(
				<Overlay show={false} placement="left">
					<Test />
				</Overlay>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when the child has className', () => {
			const wrapper = shallow(
				<Overlay show={true} placement="left">
					<Test className="TestClass" />
				</Overlay>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('sets visible to false on transition end when show is false', () => {
			const wrapper = mount(
				<Overlay show={true} placement="left">
					<Test id="child" />
				</Overlay>
			);
			wrapper.setState({visible: true});
			wrapper.setProps({show: false});
			jest.runOnlyPendingTimers();
			const child = wrapper.find('#child');
			expect(wrapper.state('visible')).toBe(false);
			expect(child.getDOMNode().classList.contains('Overlay--visible')).toBe(false);
		});

		it('updates the position when placement is left and show changes to true', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="left" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({positionLeft: 80, positionTop: 120, arrowOffsetTop: '50%', visible: true});
		});

		it('updates the position when placement is left and show changes to true and the overlay position is negative', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 400})});
			const wrapper = mount(
				<Overlay show={false} placement="left" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 80,
				positionTop: 0,
				arrowOffsetTop: '31.25%',
				visible: true,
			});
		});

		it('updates the position when placement is left and show changes to true and the overlay position exceeds the window height', () => {
			window.innerHeight = 600;
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 500, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 400})});
			const wrapper = mount(
				<Overlay show={false} placement="left" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 80,
				positionTop: 200,
				arrowOffsetTop: '81.25%',
				visible: true,
			});
		});

		it('updates the position when placement is right and show changes to true', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="right" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 300,
				positionTop: 120,
				arrowOffsetTop: '50%',
				visible: true,
			});
		});

		it('updates the position when placement is top and show changes to true', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="top" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 190,
				positionTop: 90,
				arrowOffsetLeft: '50%',
				visible: true,
			});
		});

		it('updates the position when placement is top and show changes to true and the overlay position is negative', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 800, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="top" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({positionLeft: 0, positionTop: 90, arrowOffsetLeft: '25%', visible: true});
		});

		it('updates the position when placement is top and show changes to true and the overlay position exceeds the window width', () => {
			window.innerWidth = 1000;
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 700, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 800, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="top" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 200,
				positionTop: 90,
				arrowOffsetLeft: '75%',
				visible: true,
			});
		});

		it('updates the position when placement is bottom and show changes to true', () => {
			jest
				.spyOn(ReactDOM, 'findDOMNode')
				.mockReturnValueOnce({getBoundingClientRect: () => ({top: 100, left: 100, width: 200, height: 50})})
				.mockReturnValueOnce({getBoundingClientRect: () => ({width: 20, height: 10})});
			const wrapper = mount(
				<Overlay show={false} placement="bottom" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.setProps({show: true});
			expect(wrapper.state()).toEqual({
				positionLeft: 190,
				positionTop: 150,
				arrowOffsetLeft: '50%',
				visible: true,
			});
		});

		it('does not change state when visible is true and other property changes', () => {
			const wrapper = mount(
				<Overlay show={true} placement="left">
					<Test id="child" />
				</Overlay>
			);
			wrapper.setState({visible: true});
			wrapper.setProps({other: true});
			const child = wrapper.find('#child');
			expect(child.getDOMNode().classList.contains('Overlay--visible')).toBe(true);
		});

		it('clears the timeout on unmount if set', () => {
			const wrapper = mount(
				<Overlay show={false} placement="bottom" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.instance().timer = 1;
			wrapper.unmount();
			expect(clearTimeout.mock.calls).toEqual([[1]]);
		});

		it('does not clear the timeout on unmount if not set', () => {
			const wrapper = mount(
				<Overlay show={false} placement="bottom" target={<div />}>
					<Test />
				</Overlay>
			);
			wrapper.unmount();
			expect(clearTimeout).not.toHaveBeenCalled();
		});
	});
});
