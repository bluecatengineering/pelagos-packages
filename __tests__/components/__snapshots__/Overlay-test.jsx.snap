// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay rendering renders expected elements 1`] = `
<UNDEFINED
  containerInfo={<body />}
>
  <Test
    arrowOffsetLeft={null}
    arrowOffsetTop={null}
    className="Overlay"
    placement="left"
    positionLeft={0}
    positionTop={0}
  />
</UNDEFINED>
`;

exports[`Overlay rendering renders expected elements when show is false 1`] = `null`;

exports[`Overlay rendering renders expected elements when the child has className 1`] = `
<UNDEFINED
  containerInfo={<body />}
>
  <Test
    arrowOffsetLeft={null}
    arrowOffsetTop={null}
    className="TestClass Overlay"
    placement="left"
    positionLeft={0}
    positionTop={0}
  />
</UNDEFINED>
`;
